### DISCUSSION TITLE
C++ 中的中文编码 | SF-Zhou's Blog

### DISCUSSION THREAD

**Discussion by SF-Zhou** [SF-Zhou/sf-zhou.github.io#179] [Announcements]
https://sf-zhou.github.io/programming/chinese_encoding.html

**Comment by smilethensee** [comment]
一：26 个小写字母（从 97 开始）
五：“UTF-8 编码的单个汉字长度为 3，英文长度为 1”，单个汉字不是2个字节吗？程序的输出也是两个字节。

谢谢博主！！！

**Comment by SF-Zhou** [comment]
> @smilethensee
> 一：26 个小写字母（从 97 开始）
> 五：“UTF-8 编码的单个汉字长度为 3，英文长度为 1”，单个汉字不是2个字节吗？程序的输出也是两个字节。
> 
> 谢谢博主！！！

感谢指出 typo~

UTF-8 里大部分的汉字需要三个字节来编码的，少数生僻字需要更多的字节。可以用 Python 简单验证：
```python
print(len(u'中国人'.encode('utf-8')))  # output 9
```

**Comment by smilethensee** [comment]
Got it.
我用的DEV C++, 应该不是UTF-8编码格式。在python里使用博主的代码，确实是三个字节。谢谢博主，涨姿势了。

**Comment by zchrissirhcz** [comment]
你好，我在Windows10下，VS2017里运行了你最后贴出的“最佳实践”代码。得到的test_02.txt文件，用Notepad打开显示乱码（状态栏显示UTF-8）。另存为“ANSI”后再打开，中文显示正确，并且状态栏显示UTF-8。

**Comment by SF-Zhou** [comment]
> @zchrissirhcz
> 你好，我在Windows10下，VS2017里运行了你最后贴出的“最佳实践”代码。得到的test_02.txt文件，用Notepad打开显示乱码（状态栏显示UTF-8）。另存为“ANSI”后再打开，中文显示正确，并且状态栏显示UTF-8。

你好。我自己暂时没有 Windows 的测试环境，没法复现你的问题。不过你可以检查下源代码文件本身的编码，看下是不是 UTF-8；然后对于输出的文件，可以使用 16 进制编辑器查看下内容，期望的输出应该是："e6 88 91 e7 88 b1 43 2b 2b 0a"

**Comment by zchrissirhcz** [comment]
> > @zchrissirhcz
> > 你好，我在Windows10下，VS2017里运行了你最后贴出的“最佳实践”代码。得到的test_02.txt文件，用Notepad打开显示乱码（状态栏显示UTF-8）。另存为“ANSI”后再打开，中文显示正确，并且状态栏显示UTF-8。
> 
> 你好。我自己暂时没有 Windows 的测试环境，没法复现你的问题。不过你可以检查下源代码文件本身的编码，看下是不是 UTF-8；然后对于输出的文件，可以使用 16 进制编辑器查看下内容，期望的输出应该是："e6 88 91 e7 88 b1 43 2b 2b 0a"

感谢回复。我检查了源代码文件编码，是UTF-8格式。输出文件，我用HxD64.exe查看，内容是

>E9 8E B4 E6 88 A0 E5 9F 8D 43 2B 2B 0D 0A

![c++_codec](https://user-images.githubusercontent.com/3831847/79566872-d6554500-80e5-11ea-9d37-74b1d2ad99f7.png)

**Comment by SF-Zhou** [comment]
> @zchrissirhcz
> > > @zchrissirhcz
> > > 你好，我在Windows10下，VS2017里运行了你最后贴出的“最佳实践”代码。得到的test_02.txt文件，用Notepad打开显示乱码（状态栏显示UTF-8）。另存为“ANSI”后再打开，中文显示正确，并且状态栏显示UTF-8。
> > 
> > 你好。我自己暂时没有 Windows 的测试环境，没法复现你的问题。不过你可以检查下源代码文件本身的编码，看下是不是 UTF-8；然后对于输出的文件，可以使用 16 进制编辑器查看下内容，期望的输出应该是："e6 88 91 e7 88 b1 43 2b 2b 0a"
> 
> 感谢回复。我检查了源代码文件编码，是UTF-8格式。输出文件，我用HxD64.exe查看，内容是
> 
> >E9 8E B4 E6 88 A0 E5 9F 8D 43 2B 2B 0D 0A
> 
> ![c++_codec](https://user-images.githubusercontent.com/3831847/79566872-d6554500-80e5-11ea-9d37-74b1d2ad99f7.png)
> 

测试了一下，VS 比较迷，代码文件需要使用 ANSI 编码才会有期望的输出。所以现在直接在字符串里给了 Unicode 值了，测试就 OK 了。

**Comment by 1blezzed1** [comment]
hello
